refactor: publish error handling

yndajas opened the pull request:
**Problem:** going through the form and clicking on publish returns errors through JSON. Currently we're just throwing these errors into the errors banner without highlighting the missing required fields in red. The method we use to add the red highlight on the edit form involves adding a class to the missing required fields by using a SurveyJS hook that only fires on rendering questions

**Solution:** we could probably hack things and add in the classes with some funky JavaScript without having to save an unpublishable commitment, but that might be complicated. Instead, I've made changes to the controller so that if a user tries to publish an unpublishable commitment, it saves the commitment as draft and sends back a redirect path of the edit form. The Vue component then redirects to the edit form and the lifecycle method we're using to add classes to the missing required fields kicks in

**Update:** the error banner might be reworked a little to include a close button, but the flow implemented in this PR is ready for review

yndajas commented on app/controllers/commitments_controller.rb:
```diff
@@ -118,6 +103,53 @@ def destroy
   private
 
+  def publish_or_save
+    @is_new_record = @commitment.new_record?
```
Think you're right - probably was a bit overengineered, and some of the methods didn't actually share any logic for create and update, so no need to abstract them. Have updated

yndajas commented:
Merge #104 into this and test everything together before merging into `develop`

sergiomarrocoli commented on app/models/commitment.rb:
```diff
@@ -34,7 +34,9 @@ class Commitment < ApplicationRecord
+  accepts_nested_attributes_for :progress_documents, reject_if: ->(attributes) {
+    attributes['document'].blank? && attributes['_destroy'].blank?
+  }, allow_destroy: true
 
```
validates :latitude , numericality: { greater_than_or_equal_to:  -90, less_than_or_equal_to:  90 }, if: :user_created_and_live?
  validates :longitude, numericality: { greater_than_or_equal_to: -180, less_than_or_equal_to: 180 }, if: :user_created_and_live?

sergiomarrocoli reviewed (commented):
i think the errors are all handled correctly, but they sometimes (always?) only show if you navigate away from the page and back (for back end validation errors) or try to move forward (surveyjs errors). Did you mentioned we might be able to activate these without doing that?

yndajas commented:
@sergiomarrocoli made a few updates:

- the page is now changed and changed back to trigger immediate error CSS class setting

- error highlighting now works for non-presence-related validations - previously the red highlight was only there if the question was unanswered (i.e. no value), so it only worked for missing required fields. You can confirm this by checking if lat/long validation errors are highlighted in the question title (not sure if any others can be checked as SurveyJS should be preventing them)

- changed latitude/longitude to number inputs, meaning no non-numeric characters (except -/+) can be entered. I tried including a max/min but when you do this you can still type an out of range number, and SurveyJS doesn't immediately validate on changing the value so you can hit submit - SurveyJS then strips the invalid data rather than sending it as is to the backend, so you get redirected to the dashboard with no data change for that field but also no error message. Without a max/min it'll always submit the data, so the server validations can provide an error message - more user friendly

- lat/long validations are in place

sergiomarrocoli commented on test/controllers/commitments_controller_test.rb:
```diff
@@ -76,7 +76,7 @@ class CommitmentsControllerTest < ActionDispatch::IntegrationTest
 
-  test 'should return errors if attempt to create a live commitment fails' do
+  test 'should save commitment as draft if attempt to create a live commitment fails' do
     sign_in users(:user_1)
```
line 57 failing, because of new error structure. should be: ```assert JSON.parse(response.body)['errors'].include?('name')```

sergiomarrocoli commented on app/models/commitment.rb:
```diff
@@ -225,6 +238,19 @@ def self.pages(item_count)
   end
 
+  def progress_documents_attributes=(attrs)
+    attrs.each do |attr|
```
this can be removed

sergiomarrocoli commented on app/models/commitment.rb:
```diff
@@ -34,7 +34,9 @@ class Commitment < ApplicationRecord
 
   accepts_nested_attributes_for :links, reject_if: ->(attributes) { attributes['url'].blank? }, allow_destroy: true
-  accepts_nested_attributes_for :progress_documents, reject_if: ->(attributes) { attributes['document'].blank? }, allow_destroy: true
+  accepts_nested_attributes_for :progress_documents, reject_if: ->(attributes) {
```
i don't think these reject_if work (maybe reject_if doesn't work at all for has_many?), and now you have the front end submitting the params correctly we don't need them, so can be replaced with: ```  accepts_nested_attributes_for :links, allow_destroy: true
  accepts_nested_attributes_for :progress_documents, allow_destroy: true```

sergiomarrocoli commented on app/javascript/components/banners/ErrorBanner.vue:
```diff
@@ -2,35 +2,21 @@
 </template>
 
-<script>
+<script>      
```
whitespace

sergiomarrocoli commented on app/javascript/components/banners/ErrorBanner.vue:
```diff
@@ -2,35 +2,21 @@
 <template>
```
can we lose the .vue on line 1?

sergiomarrocoli reviewed (changes requested):
couple of superficial comments, but otherwise it's all there. working perfectly as far as i can tell and no more param filtering needed in the back end

yndajas commented:
All done!
